lib/VolSurface/Calibration/Equities.pm

```diff
@@ -0,0 +1,648 @@
+}
+
+sub _calculate_variance {
+    my ($self, $tiy, $params) = @_;
+
+    my $sig_small  = $params->{atmvolshort}**2;
+    my $sig_1y     = $params->{atmvol1year}**2;
+    my $sig_inf    = $params->{atmvolLong}**2;
+    my $t_small    = 1.0;
+    my $atm_alpha1 = $params->{atmWingL};
+    my $atm_alpha2 = $params->{atmWingR};
+
+    my $atm_vol =
+        $sig_inf +
+        ($sig_1y - $sig_inf) *
+        ((exp(-$atm_alpha2 * $tiy) - exp(-$atm_alpha1 * $tiy)) / (exp(-$atm_alpha2 * $t_small) - exp(-$atm_alpha1 * $t_small))) +
+        ($sig_small - $sig_inf) *
+        ((exp(-$atm_alpha2 * $t_small - $atm_alpha1 * $tiy) - exp(-$atm_alpha1 * $t_small - $atm_alpha2 * $tiy)) /
+            (exp(-$atm_alpha2 * $t_small) - exp(-$atm_alpha1 * $t_small)));
+
+    return $atm_vol;
+}
+
+=head2 _optimize
+
+Algorithm change - now based on centroid calculations
+A function that optimizes a set of parameters against a function.
+This optimization method is based on Amoeba optimization. We use a form of the 
+Downhill Simplex Method or Nelder-Mead (available as the R function optim). 
+This can also be coded in other languages. 
+
+=cut
+
+sub _optimize {
+    my ($self, $params) = @_;
+
+    my $intol         = $self->_tolerance_level;
+    my $num_of_var    = scalar(@$params);
+    my $num_of_points = $num_of_var;
+
+    my $highest_value;
+
+    my $lambda1 = ((($num_of_points + 1)**0.5) - 1 + $num_of_points) / ($num_of_points * (2**0.5)) * $self->_scale;
+    my $lambda2 = ((($num_of_points + 1)**0.5) - 1) / ($num_of_points * (2**0.5)) * $self->_scale;
+
+    my $i;
+    my $j;
+
+    #initialize the simplex with the initial guess
+    my @simplex = ($params);
+
+    my $step = 0;
+    for ($i = 0; $i < $num_of_var; $i++) {
+        $step = 0.1 * abs($params->[$i])
+            if (0.1 * abs($params->[$i]) > $step);
+    }
+
+    for (my $i = 0; $i < $num_of_points; $i++) {
+        push @simplex, [map { $simplex[0][$_] + ($i == $_ ? $step : 0) } (0 .. $num_of_var - 1)];
+    }
+
+    my @function_eval = map { $self->function_to_optimize($_) } @simplex;
+    my $starting_value = $function_eval[0];
+
+    my @current_reflection;
+    my @current_expansion;
+    my @current_contraction;
+
+    my $expansion_function_eval;
+    my $reflected_function_eval;
+    my $contraction_function_eval;
+    my $lowest_point_index = 0;
+
+    my $calcvert = 1;
+    my $counter  = 0;
+
+    while (1) {
+        if ($calcvert == 1) {
+            $calcvert = 0;
+            for ($j = 0; $j <= $num_of_points; $j++) {
+                if ($j != $lowest_point_index) {
+                    $function_eval[$j] = $self->function_to_optimize($simplex[$j]);
+                }
+            }
+        }
+
+        my @sorted_eval = sort { $a <=> $b } @function_eval;
+        $lowest_point_index = first_index { $_ == $sorted_eval[0] } @function_eval;
+        my $highest_point_index = first_index { $_ == $sorted_eval[-1] } @function_eval;
```

Comment:
Ok, since you insist I measured it.

Here is my test program:

```perl
#!/usr/bin/perl

use strict;
use warnings;

use Benchmark qw/:hireswallclock cmpthese timethese/;
use Test::More;

my @numbers = map {rand} 1..(shift||7);
push @numbers, @numbers;        # to have duplicates

sub simple {
    my $l = 0;
    my $h = 0;
    for (my $i=1; $i<@numbers; $i++) {
        $l = $i if $numbers[$i] < $numbers[$l];
        $h = $i if $numbers[$i] > $numbers[$h];
    }
    return $l, $h;
}

sub complex {
    use List::MoreUtils qw/first_index/;

    my @sorted = sort {$a<=>$b} @numbers;
    my $l = first_index {$_ == $sorted[ 0]} @numbers;
    my $h = first_index {$_ == $sorted[-1]} @numbers;
    return $l, $h;
}

my @simple  = simple;
my @complex = complex;

# both, simple and complex, are stable. That means the
# highest and lowest indices must be in the first half.

cmp_ok $simple[0], '<', @numbers/2, 'simple: l<'.(@numbers/2);
cmp_ok $simple[1], '<', @numbers/2, 'simple: h<'.(@numbers/2);

cmp_ok $complex[0], '<', @numbers/2, 'complex: l<'.(@numbers/2);
cmp_ok $complex[1], '<', @numbers/2, 'complex: h<'.(@numbers/2);

{
    use List::Util qw/max min/;
    is $numbers[$simple[0]], min(@numbers), 'simple: l is correct';
    is $numbers[$simple[1]], max(@numbers), 'simple: h is correct';
    is $numbers[$complex[0]], min(@numbers), 'complex: l is correct';
    is $numbers[$complex[1]], max(@numbers), 'complex: h is correct';
}

done_testing;

cmpthese timethese shift||-10, {
    simple  => \&simple,
    complex => \&complex,
};
```

And here is the result:

```
0 [09:49] nobody@qa18 /home/git$ perl xx.pl
ok 1 - simple: l<7
ok 2 - simple: h<7
ok 3 - complex: l<7
ok 4 - complex: h<7
ok 5 - simple: l is correct
ok 6 - simple: h is correct
ok 7 - complex: l is correct
ok 8 - complex: h is correct
1..8
Benchmark: running complex, simple for at least 10 CPU seconds...
   complex: 10.1257 wallclock secs (10.05 usr +  0.00 sys = 10.05 CPU) @ 187964.78/s (n=1889046)
    simple: 10.1202 wallclock secs (10.11 usr +  0.00 sys = 10.11 CPU) @ 257753.41/s (n=2605887)
            Rate complex  simple
complex 187965/s      --    -27%
simple  257753/s     37%      --
```

I ran it several times and `simple` was something between 25% and 60% faster than `complex`. This is less improvement than I would have expected. Also, there is an interesting thing going on. For small lists like in our case (just 13 numbers), `simple` always wins. When the list gets longer, `complex` wins. For lists of 100 to 1000 elements, `complex` is on my qa box up to 60% faster. But when the lists get even longer, `simple` catches up again. On my notebook, `simple` started to win again for lists of around 1500 elements. On the qa box, it's almost even. For really long lists (~50k elements) `simple` always wins again even on the qa box.

So, given that this module isn't used for every pricing attempt and given that the effect is relatively small, I am fine with your solution.